### PR TITLE
Add analyzer batching

### PR DESCRIPTION
**Pull request by shahrukhx01**
@lalitpagaria  I have added the batching on `ZeroShotAnalyzer` to get your feedback on changes that I made and design in general, I used the following code to benchmark performance, seems like batch size of `1` are outperforming larger batches which is counter-intuitive this may change on `GPU` I have only tested on local `CPU`

```python
if __name__ == "__main__":
    import timeit  ## for benchmarking execution

    start = timeit.default_timer()
    GOOD_TEXT = """If anyone is interested... these are our hosts. I can’t recommend them enough, Abc & Pbc.

    The unit is just lovely, you go to sleep & wake up to this incredible place, and you have use of a Weber grill and a ridiculously indulgent hot-tub under the stars"""

    BAD_TEXT = """I had the worst experience ever with XYZ in Egypt. Bad Cars, asking to pay in cash,  do not have enough fuel,  do not open AC,  wait far away from my location until the trip is cancelled,  call and ask about the destination then cancel, and more. Worst service."""

    MIXED_TEXT = """I am mixed"""

    TEXTS = [
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
        GOOD_TEXT,
        BAD_TEXT,
        MIXED_TEXT,
    ] * 2
    zero_shot_analyzer = ZeroShotClassificationAnalyzer(
        model_name_or_path="typeform/mobilebert-uncased-mnli",
    )
    batch_size = 4
    labels = ["facility", "food", "comfortable", "positive", "negative"]

    source_responses = [
        AnalyzerRequest(processed_text=text, source_name="sample") for text in TEXTS
    ]
    analyzer_responses = zero_shot_analyzer.analyze_input(
        source_response_list=source_responses,
        analyzer_config=ClassificationAnalyzerConfig(
            labels=labels, batch_size=batch_size
        ),
    )
    stop = timeit.default_timer()

    print("Time: ", stop - start)
    print("Batch size: ", batch_size)
    print("Total Texts: ", len(TEXTS))
    """for res in analyzer_responses:
        print(res.segmented_data)
    """
    assert len(analyzer_responses) == len(TEXTS)

    for analyzer_response in analyzer_responses:
        assert len(analyzer_response.segmented_data) == len(labels)
        assert "positive" in analyzer_response.segmented_data
        assert "negative" in analyzer_response.segmented_data

```

### CODE REVIEW

**Comment by lalitpagaria** [comment]
@shahrukhx01 design wise your changes are good. It is nice that you have used Generator.
Regarding performance for batch_size > 1, can you please check on colab with GPU env?

It seems few transformers pipeline already support batching, we can relay on their strategy for batching. In this case we can pass List[str] to transformers pipeline.

**Comment by shahrukhx01** [comment]
@lalitpagaria Batching on does make the inference time half. See here:
[Analyzer Benchmarking Colab Notebook](https://colab.research.google.com/drive/11DOwi6HyfCNMxm2KiVxWo9k-pdYSaRjb?usp=sharing)

**Comment by lalitpagaria** [comment]
Nice. Can you please add one more test to your benchmark. Generate text randomly. Just to rule out any intermediate catching.

File: obsei/analyzer/classification_analyzer.py [hunk 1]
```diff
@@ -34,25 +34,42 @@ def __init__(self, **data: Any):
...
-        self, text: str, labels: List[str], multi_class_classification: bool = True
-    ) -> Dict[str, float]:
-        scores_data = self._pipeline(
-            text[: self._max_length] if len(text) > self._max_length else text,
+        self,
+        texts: List[str],
+        labels: List[str],
+        multi_class_classification: bool = True,
+    ) -> List[Dict[str, float]]:
+        if len(texts) <= 1:
```
Comment (lalitpagaria) [inline obsei/analyzer/classification_analyzer.py:48]: Why this condition? it seems it is not required

**Comment by lalitpagaria** [comment]
Just one small comment otherwise looks fine to me.

**Comment by shahrukhx01** [comment]
> Nice. Can you please add one more test to your benchmark. Generate text randomly. Just to rule out any intermediate catching.

@lalitpagaria generating a new batch of texts for each text gives a slower performance, however, it is approx. 2x faster than using batch of size=1
[Colab Notebook](https://colab.research.google.com/drive/11DOwi6HyfCNMxm2KiVxWo9k-pdYSaRjb?usp=sharing)

**Comment by lalitpagaria** [comment]
All fine. Please add to other analyzers

**Comment by shahrukhx01** [comment]
@lalitpagaria CI would fail on some instances because some test cases are based on single instance inference. i.e., ner analyzer etc

**Comment by shahrukhx01** [comment]
Update: I have added batching to all transformers-based analyzers, for Vader and Presidio-based analyzers I'll have to double-check whether the batching is supported and what is the optimal way doing it there.

**Review by shahrukhx01 — COMMENTED** [review]
> Update: I have added batching to all transformers-based analyzers, for Vader and Presidio-based analyzers I'll have to double-check whether the batching is supported and what is the optimal way doing it there.

Vader doesn't support batch call:

```python
def polarity_scores(self, text):
        """
        Return a float for sentiment strength based on the input text.
        Positive values are positive valence, negative value are negative
        valence.
        """
        # convert emojis to their textual descriptions
        text_no_emoji = ""
        prev_space = True
        for chr in text:
            if chr in self.emojis:
                # get the textual description
                description = self.emojis[chr]
                if not prev_space:
                    text_no_emoji += ' '
                text_no_emoji += description
                prev_space = False
            else:
                text_no_emoji += chr
                prev_space = chr == ' '
        text = text_no_emoji.strip()

        sentitext = SentiText(text)
```

**Comment by shahrukhx01** [comment]
> Update: I have added batching to all transformers-based analyzers, for Vader and Presidio-based analyzers I'll have to double-check whether the batching is supported and what is the optimal way doing it there.

Presidio doesn't support batch call:
Analyzer
```python
def analyze(
        self,
        text: str,
        language: str,
        entities: Optional[List[str]] = None,
        correlation_id: Optional[str] = None,
        score_threshold: Optional[float] = None,
        return_decision_process: Optional[bool] = False,
        ad_hoc_recognizers: Optional[List[EntityRecognizer]] = None,
    ) -> List[RecognizerResult]:
```
Anonymizer
```python
def anonymize(
            self,
            text: str,
            analyzer_results: List[RecognizerResult],
            operators: Optional[Dict[str, OperatorConfig]] = None
    ) -> EngineResult:
```

**Comment by shahrukhx01** [comment]
@lalitpagaria I have updated the test cases and added batching to all transformer analyzers for others it's not possible as those libraries expect one text at a time. Please review and let me know if any change is needed.

File: obsei/analyzer/classification_analyzer.py (hunk 1 above)
Comment (lalitpagaria) [reply obsei/analyzer/classification_analyzer.py:48]: How about this then -

```
prediction = self._pipeline(
                    texts,
                    labels,
                    multi_label=multi_class_classification
)
return prediction if isinstance(prediction, list) else [prediction]
```

File: obsei/analyzer/base_analyzer.py [hunk 2]
```diff
@@ -30,6 +30,7 @@ def to_dict(self):
 
 class BaseAnalyzerConfig(BaseModel):
     TYPE: str = "Base"
+    batch_size: int = 1
```
Comment (lalitpagaria) [inline obsei/analyzer/base_analyzer.py:33]: Set default value to 64

Comment (shahrukhx01) [reply obsei/analyzer/base_analyzer.py]: @lalitpagaria done.

File: obsei/analyzer/translation_analyzer.py [hunk 3]
```diff
@@ -19,26 +25,55 @@ def __init__(self, **data: Any):
...
+            yield (
+                texts[index : index + batch_size],
+                source_response_list[index : index + batch_size],
+            )
+
     def analyze_input(
-        self, source_response_list: List[AnalyzerRequest], **kwargs
+        self,
+        source_response_list: List[AnalyzerRequest],
+        analyzer_config: BaseAnalyzerConfig,
```
Comment (lalitpagaria) [inline obsei/analyzer/translation_analyzer.py:48]: Can you please run mypy, it will it seems type checking will fail some where.

**Comment by lalitpagaria** [comment]
Overall PR is good. Thank for working on it.
I have few minor comments.

File: obsei/analyzer/translation_analyzer.py (hunk 3 above)
Comment (shahrukhx01) [reply obsei/analyzer/translation_analyzer.py:48]: @lalitpagaria 
```bash
mypy obsei/analyzer/translation_analyzer.py --ignore-missing-imports
```

response:
Success: no issues found in 1 source file

File: obsei/analyzer/classification_analyzer.py (hunk 1 above)
Comment (lalitpagaria) [reply obsei/analyzer/classification_analyzer.py:48]: @shahrukhx01 I think this one last pending :)

Comment (shahrukhx01) [reply obsei/analyzer/classification_analyzer.py]: @lalitpagaria I missed this one. Done now!

**Comment by lalitpagaria** [comment]
@shahrukhx01 CI is failing https://github.com/lalitpagaria/obsei/pull/118/checks?check_run_id=2737130061

**Comment by shahrukhx01** [comment]
> @shahrukhx01 CI is failing https://github.com/lalitpagaria/obsei/pull/118/checks?check_run_id=2737130061

happening because of the last change, let me take a look.

**Comment by shahrukhx01** [comment]
> > @shahrukhx01 CI is failing https://github.com/lalitpagaria/obsei/pull/118/checks?check_run_id=2737130061
> 
> happening because of the last change, let me take a look.

@lalitpagaria fixed, return type for classification pipeline of NER is `List[List[Dict]]`. Everything is good to go now!

**Comment by lalitpagaria** [comment]
@shahrukhx01  Thanks for the PR.

**Comment by shahrukhx01** [comment]
@lalitpagaria you are welcome! :)

**Pull request merged**